### PR TITLE
Full Implementation of Recycle Bin Feature and Updates to Tests

### PR DESCRIPTION
**Pull request by case141**
Completed implementation of Recycle Bin Functionalities! Recycle Bin is now a mode that users can switch to and from the usual Source Manager mode. It reads data from the deletesource.json database. You can only use certain commands specified under the Recycle Bin section in the UserGuide for this mode.

**1. Updates to parser structure**
- create RecycleBinParser class
- made SourceManagerParser a parent of RecycleBinParser

**2. Updates to LogicManager**
- implemented a switch statement to switch between RecycleBinParser and SourceManagerParser

**3. Updates to Model Component**
- added an Enum class called ParserMode that contains `RECYCLE_BIN` and `SOURCE_MANAGER`
- add getters and setters for ParserMode

**4. Commands changed/introduced**
- `recycle-bin`: previously the `list-deleted` feature, now other than listing the deleted sources, it also switches parser modes from RecycleBinParser to SourceManagerParser and vice versa
- `restore`: only works in Recycle Bin mode now
- `delete`: on Source Manager - non-permanent delete, on Recycle Bin - permanent delete
- `exit-bin`: exits recycle bin and switches back to source manager
- `empty-bin`: clears only the recycle bin

**5. Uncommented All SystemTests and add @ignore instead, they have to be fixed**
image

**6. Fixed all tests except the following**
image

**7. Change Sample Test Data to suit Infinity Machine**
- Replaced the `ALICE, BOB, etc.` to `ALGORITHM_RESEARCH, etc.` in TypicalSources.java to suit Infinity Machine 

**Tasks left:**
- Update UserGuide and DeveloperGuide with Recycle Bin details
- Add diagram for Recycle Bin Structure
- Fix SystemTests and Tests
- Add Tests for Recycle Bin 

Implements #210 #153

### CODE REVIEW

File: src/main/java/seedu/address/commons/core/Messages.java [hunk 1]
```diff
@@ -8,6 +8,8 @@
     public static final String MESSAGE_UNKNOWN_COMMAND = "Unknown command.";
     public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! \n%1$s";
     public static final String MESSAGE_INVALID_SOURCE_DISPLAYED_INDEX = "The source index provided is invalid";
+    public static final String MESSAGE_UNAVAILABLE_COMMAND = "This command is unavailable in this mode,\n"
```
Comment (fterh) [inline src/main/java/seedu/address/commons/core/Messages.java]: Weird formatting. Why is the space after `\n`?

File: src/main/java/seedu/address/logic/parser/SourceManagerParser.java [hunk 2]
```diff
@@ -36,23 +35,23 @@
...
 /**
- * Parses user input.
+ * Parses user input for Source Manager.
  */
 public class SourceManagerParser implements CommandValidator {
-    private static final String COMMAND_ALIAS_ADD = "alias";
-    private static final String COMMAND_ALIAS_REMOVE = "alias-rm";
-    private static final String COMMAND_ALIAS_CLEAR = "alias-clear";
-    private static final String COMMAND_ALIAS_LIST = "alias-ls";
+    protected static final String COMMAND_ALIAS_ADD = "alias";
```
Comment (fterh) [inline src/main/java/seedu/address/logic/parser/SourceManagerParser.java]: This doesn't have to be protected anymore, right? Since recycle bin doesn't support aliasing anymore.

File: src/main/java/seedu/address/logic/commands/UndoCommand.java [hunk 3]
```diff
@@ -24,13 +24,12 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
             throw new CommandException(MESSAGE_FAILURE);
         }
 
-        if (!model.canUndoSourceManager()) {
+        if (!model.canUndoDeletedSources()) {
```
Comment (fterh) [inline src/main/java/seedu/address/logic/commands/UndoCommand.java:27]: Why did you change this line?

File: src/main/java/seedu/address/logic/parser/SourceManagerParser.java [hunk 4]
```diff
@@ -36,23 +35,23 @@
...
 
     /**
      * Used for initial separation of command word and args.
      */
-    private static final Pattern BASIC_COMMAND_FORMAT = Pattern.compile("(?<commandWord>\\S+)(?<arguments>.*)");
+    protected static final Pattern BASIC_COMMAND_FORMAT = Pattern.compile("(?<commandWord>\\S+)(?<arguments>.*)");
 
-    private Set<String> metaCommands = new HashSet<>();
-    private Set<String> validCommands = new HashSet<>();
+    protected Set<String> metaCommands = new HashSet<>();
```
Comment (fterh) [inline src/main/java/seedu/address/logic/parser/SourceManagerParser.java]: Same here

File: src/main/java/seedu/address/model/ModelManager.java [hunk 5]
```diff
@@ -435,9 +436,24 @@ public void setNumberOfPinnedSources(int newNumber) {
         this.numPinnedSources = newNumber;
     }
 
+    @Override
+    public void setParserMode(ParserMode mode) {
+        if (mode == ParserMode.SOURCE_MANAGER) {
```
Comment (fterh) [inline src/main/java/seedu/address/model/ModelManager.java]: Use switch case instead

File: src/main/java/seedu/address/logic/parser/SourceManagerParser.java (hunk 2 above)
Comment (case141) [reply src/main/java/seedu/address/logic/parser/SourceManagerParser.java]: right i forgot to change that thanks for pointing out :)

File: src/main/java/seedu/address/logic/commands/UndoCommand.java (hunk 3 above)
Comment (case141) [reply src/main/java/seedu/address/logic/commands/UndoCommand.java:27]: previously it was 2 duplicate if statements bcos the undo commands needs to be applied for deletedsource.json as well so I change the duplicate statement to canUndoDeletedSources()

File: src/main/java/seedu/address/logic/parser/SourceManagerParser.java (hunk 4 above)
Comment (case141) [reply src/main/java/seedu/address/logic/parser/SourceManagerParser.java]: yup

File: src/main/java/seedu/address/model/ModelManager.java (hunk 5 above)
Comment (case141) [reply src/main/java/seedu/address/model/ModelManager.java]: good suggestion!

**Review by fterh — APPROVED** [review]
